Intoroduce io.finch.route package

vkostyukov opened the pull request:
This is an initial implementation of route combinators as discussed in #146. There is no DSL layer yet - just raw classes. So, for now we have to explicitly create new instances like `StringMatcher` and `StringExtractor`. The next step would be to write a bunch of implicit conversion in order to enable full-featured DSL.

But now I wanted to collect a feedback and ask for an advice. I don't really think we need both `RouteMatcher` and `RouteExtractor`. My primary goal now is to replace them both with just `Router`. I would love to hear your advices @pasviegas, @benjumanji, @rodrigopr and @rpless on how can we achieve it.

The main problem here is that there are routers that don't extract anything but match something: `Get`, `StringMatcher`. So, the problem is in `RouteExtractor.flatMap` method. Let's say we have just `Router` and define `Get` as `Router[Unit]`. And when we flatMap it to `Router[Int]` we will get `Router[Unit / Int]`, which is wrong.

Anyway. The main idea i was trying to show is that router is just a function `Route => A`. And if `A` is `Service[Req, Rep]` we can call such router an `Endpoint`, since it takes a route and returns a service. This is exactly what endpoint is.

vkostyukov commented:
If it's turned to be impossible we may rename things:
* `RouteExtractor[A]` -> `RouterN[A]`
* `RouteMatcher` -> `Router0`

And then create an aliases: 
```scala
type Router[A, B] = RouterN[A, B]
type Endpoint[A, B] = Router[Service[A, B]]
```
Usually we don't need write types but we need we should write them in a nice way:

```scala
val r: Router[Int / Int / String] = ???
```

vkostyukov commented:
@jenshalm it would also be nice to get your feedback on this.

rpless commented:
I would recommend not using `/` as the infix case class in the Router's parameters. It might conflict with an existing class name in the Finagle route api: https://github.com/twitter/finagle/blob/master/finagle-httpx/src/main/scala/com/twitter/finagle/httpx/path/Path.scala#L92.

vkostyukov commented:
Right. But I think that with route combinators users no longer need to import `finagle.httpx.path._` class. Why I do like `/` as the infix case class is that is looks like an HTTP path delimiter, which makes it fitting in the DSL thing really well.

vkostyukov commented:
What I also plan to do here is:

* `Long` extractor
* `Router.toString` should return a string representation of the router. Something like `GET "/users/:user"` (this has been asked recently on Gitter)
* More tests
* A DSL layer - please, go ahead and leave a comment on how do see the DSL for route combinators.

jenshalm commented:
Looks like a good start. But it's hard to judge the feature before seeing the planned DSL, as I assume that the base work available now is more intended to be internal API (if it would be public API I'd find it a bit too verbose).

Other than that two things come to mind:

* Are there plans to let the routing layer figure out 405 errors? In this case it might require to offer nesting method declarations inside path matchers

* A little thing: Like in some other places in Finch there is the change to reduce boilerplate, e.g. through creating a base extractor like this:

```
  class Extractor[A](f: String => Option[A]) extends RouterN[Int] {
    def apply(route: Route): Option[(Route, Int)] = for {
      (r, ii) <- StringExtractor(i)(route)
      iii <- f(ii)
    } yield (r, iii)
  }
```

which would then allow concise definitions like this:

```  
  case class IntExtractor(i: String) extends Extractor(stringToSomeNumber(_.toInt))
```

(this would require to change the stringToSomeNumber signature accordingly)

vkostyukov commented:
Thanks for the feedback @jenshalm! You're right. This is an internal API, which will be covered by a DSL layer. Wha I managed to get so far is implicit conversions for string/int matchers. So, we can already write something like this:

```scala
val router = Get / "users" / 100
``

Regarding to the universal extractor. I tried to implement the similar approach to `StringToValueOrFail` from request reader abstraction. So, the [`StringToValue`](https://github.com/finagle/finch/pull/147/files#diff-013dead8f7e981221fbdc102176fa21dR252) router does nothing that tries to evaluate the given conversion. This approach worked well for request readers, so I decided just to reuse this abstraction in some way.

Regarding the `405` error. This is an interesting question. What you can do for now (having the route combinators) is have a router that represents the rest of the path. Lets say you have an endpoint: `? /users/:id`, where `?` is either `POST` or `PATH`. So, we want to response 405 in case of other methods. Note, that we still don't have a DSL for extractors only for matchers.

```scala
val rest = "users" / IntExtractor("id")
val r = ((Path | Post) / rest ~> goodService()) | (SkipOne / rest ~> respond405())
```

vkostyukov commented:
Added DSL for `SkipOne` and replaced `SkipAll` with the end of the route `$`. See [here](https://github.com/finagle/finch/pull/147/files#diff-15fab5bf685af131056294bdbbcd0950R103) for usage details.

jenshalm commented:
I'm afraid we are talking at cross purposes regarding the generic extractor. I am not suggesting any changes for the API of the existing extractors, but only two things: a) avoid copy-paste coding as found in the RequestReaders and Extractors (once you add a LongExtractor similar to the IntExtractor) and b) as a side-effect allow users to define custom RequestReaders/Extractors in a concise API. What you wrote somehow does not sound like a counter-argument to my suggestion.

vkostyukov commented:
I think this is arguable that users will end up writing their own extractors. There is not to much information  is usually passed via URI. 

And I don't really think users write their own request readers. The `RequestReader` abstraction allows to compose readers together using the monadic API and get any readers we might want, which is a functional way to compose things but not inherit them. But I totally agree with you about the boilerplate code that exists in both packages. I'm just saying that  if user want to extend `RequestReader` to write hist own one, he is probably doing something wrong.

In case of `Router` I decide to add a `maybeMap` function into `RouterN`, which looks as follows:
```scala
    def maybeMap[B](fn: A => Option[B]): RouterN[B] = new RouterN[B] {
      def apply(route: Route): Option[(Route, B)] = for {
        (r, a) <- self(route)
        b <- fn(a)
      } yield (r, b)
    }
```

So, we can define `Extractor` and `IntExtractor` as follows:

```scala
  class Extractor[A](name: String, f: String => Option[A]) extends RouterN[A] {
    def apply(route: Route): Option[(Route, A)] = StringExtractor(name).maybeMap(f)(route)
  }
  case class IntExtractor(name: String) extends Extractor(name, stringToSomeValue(_.toInt))
```

BenWhitehead commented:
Here is another scala library that provides a DSL for URI's perhaps some of the things it does could be leveraged here, or provide guidance.

https://github.com/NET-A-PORTER/scala-uri

vkostyukov commented:
Thanks @BenWhitehead for sharing! Meanwhile, what do you guys thing about the following DSL for extractors:

```scala
val router = Get / "users" / "id" @ int / "name" @ string
```

jenshalm commented:
Thanks for your explanations. I agree regarding preferring composition over inheritance. However I wasn't necessarily thinking of something that would require inheritance, but more of a lack of flexibility I see in the current request readers, but that does not belong here, so I'll stop and maybe create another issue or PR that explains what I meant in a better way.

Regarding your example for handling a 405, that would certainly work, but for something that I would usually want for all of my routes it looks like too much plumbing, still. I think handling 405 is something I'd expect to be provided by the web stack for free.

The extractor DSL looks good, just that I am not sure I would expect to be forced to provide a name for the dynamic path elements (I think many other routing DSLs do not do that). I think having something as concise as this:

```
val router = Get / "users" / int / string
```

would be neat. For rendering the route as a string, the framework could pick defaults based on the type, e.g. `:int`. But then again, it's not a big deal if that's not possible.

vkostyukov commented:
Thanks @jenshalm! Please, go ahead with the issue/PR about request readers. I would love to understand your concern and fix it in Finch.

Regarding the 405 error. I belie most of the frameworks/libraries will respond 404 in this case. If you know libraries that support 405 out of the box please let me know. I'm pretty sure there is a way to do that using route compositor, I'm just trying to understand whether or not it worths doing this right now. I mean, we are going to cover 99% use cases of most of the users and there is also 1% that requires more changes. Alright. Let me think about this. I will try to understand the size of changes to support 405 error.

We can definitely just have `string`, `int` etc. I was just thinking about string representations. We need this for a different reasons. First of all there is a ticket #73 that will require it. And also a Finch user was asked this on Gitter recently. What I'm planing to do is to support both ways to write routers: with custom name and with default name:

```scala
val r1 = Get / "users" / int
val r2 = Get / "users" / int("id")
val r3 = Get / "users" / "id" @ int
```

The question here is what looks better `r2` or `r3`? It's definitely simpler to implement `r2`.

vkostyukov commented:
I think this 405 error handling is almost impossible to do using pure route combinators, since it will require to add a new _special_ type of combinator that handles method and when we need to support any other error we will end up adding new type/class of combinators. The reason for this is that we can't really replace `Option` with `Either` or `Try` to propagate the exception since this is not what router do and in case of router composition with `orElse` it's not really clear how to handle the exception. 

What I think we can do here is to use Finagle filters + Finch route combinators. I believe, this would be the most right way to do it. Let's say you have a filer `MakeSureMethodIs(methods: Seq[Method])` that makes sure that given request has an acceptable method and throws/responds 405 otherwise. So we can easily use it like this:

```scala
val r = * / "users" / int ~> MakeSureMethodIs(Seq(Post, Patch)) andThen editUser(_)
```

I think this is a reasonable boilerplate and more importantly we use the strong sides of both libraries: Finch and Finagle. What do you think @jenshalm? I was trying to google any frameworks that support it but found only Django that does it via decorators written by user, which is quite similar to the filters approach.

vkostyukov commented:
Enabled DSL layer for everything and rewrote all the tests using DSL:

```scala
val r1 = Get / "a" / int / string
val r2 = Get / "a" / "id" @@ int / "b" / "2" / $
```

jenshalm commented:
I think r2 and r3 work equally well in terms of being intuitive, r2 maybe just feels a tiny bit more aligned with r1?

Regarding 405 I agree in terms of design consequences potentially being severe. For that reason it is probably not amongst the most critical features for now. Regarding frameworks that support it out of the box, I think spray does (but I'm not 100% certain, this is more from reading their docs than from verifying it myself).

Regarding the PR, I'll wait a bit before I do the one for RequestReader flexibility, as first I'll be doing one for making RequestReader applicative (issue #145). I have a fully working solution here on my machine, including tests, just have to polish it a bit. (I'm currently on the same team as Ben who suggested that functionality).

vkostyukov commented:
@jenshalm this is super cool! Please, go ahead with that PR! I'm really looking forward to it.

And I just realised that it's not really possible to have both `int` and `int("name")` things in the same scope. So, I decided to go forward with `@@` syntax.

Regarding the 405 error. I will have a look on spray. Probably it will help me to understand how to integrate it in Finch. But for now, I think the Finagle filter is probably the best way to do that.

vkostyukov commented:
I had to replace `~>` operator with `/>` operator to allow us to write things w/o braces:

```scala
val r = Get / "a" /> 10 /> { _ + 20 }
```

This wont compile with `~>` since it has a higher precedence then `/`.

vkostyukov commented:
I'm pretty happy with the state of this PR. The only thing I wanted to investigate is the ability to use futures directly in the endpoint rather then services. Otherwise, I think this is ready to be shipped.

Also, please, let me know what do you guys think about `\>` alias for `map`.

benjumanji commented:
I am very wary of the operator soup that we seem to be generating. I am
traveling for the next couple of days, but I would really prefer it if we
could figure out some way to have int("a") over another operator. Surely we
can do some trick with an overloaded apply method? I won't have time to
look into it before the weekend but if this isn't merges before then I will
have a go.

On Mon, 26 Jan 2015 4:40 pm Vladimir Kostyukov <notifications@github.com>
wrote:

> I'm pretty happy with the state of this PR. The only thing I wanted to
> investigate is the ability to use futures directly in the endpoint rather
> then services. Otherwise, I think this is ready to be shipped.
>
> Also, please, let me know what do you guys think about \> alias for map.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/finagle/finch/pull/147#issuecomment-71569073>.
>

vkostyukov commented:
Thanks @benjumanji for the feedback! I will try to work out `int("name")` then. Do you have any preferences on `/>` operator? Does it look reasonable?

pasviegas commented:
Maybe it would be a good idea to move the operators to a different package, like "syntax", so the "operator soup" is optional.

vkostyukov commented:
But route combinators w/o syntax make no sense. Nobody wants to write `Get flatMap int flatMap string`. Thus, I think we should provide a reasonable set of operators for route combinator out of the box.

vkostyukov commented:
If somebody has a good candidate for `map` instead of `/>`, please let me know,

vkostyukov commented:
I turned to be super easy to replace `@@` syntax with `int("name")`.

vkostyukov commented:
Added `toString` method that converts routers to string:

```scala
val r4 = Get / string("name") / int("id") / boolean("flag") / "foo"
r4.toString shouldBe "GET /:name/:id/:flag/foo"
```

rodrigopr commented:
Cool @vkostyukov, the new API is looking good already.

I personally prefer the clarity of partial functions when describing multiple routes,
we may be able to continue using it with this api if `Extractor` had a `unapply` method.

Along with a helper function like:
```scala
object Endpoint { 
  def build[Rep, Rep](router: PartialFunction[Router[_], Service[Req, Rep]]): Endpoint[Req, Rep]
}
```

What do you think?

vkostyukov commented:
@rodrigopr Looks interesting. I think we keep the old-style `Endpoint` for some time, so you can continue using it. Once we consider deprecating `Endpoint`, we might think about the approach you suggested. But I personally think route combinators are the nicest way to do that. In case of multiple routes if might looks like (using for-comprehension):

```scala
val users = for { id / name <- Get / "users" / int / string } yield GetUser(id, name) |
            for { id <- Post / "users"} yield PostUser(id) 

val backend = users | tickets | tags
```

or using braces:

```scala
val users = (Get / "users" / int / string /> { case a / b => GetUser(id, name)}) |
            (Post / "users" / int /> PostUser(_))
```

Finch will be definitely supporting the old-school `Endpoint` but the new API will be preferred.

@rodrigopr what do you think about this syntax?

BenWhitehead commented:
I think I'm with @rodrigopr, I think I prefer the Endpoint with partial function approach from early finch.

For example I've got an existing service written in finch that looks similar to this:
```scala
  def route = {
    case Method.Get -> Root / "item" / Integer(itemId) / "recipe" / "summary" => 
      itemRecipeSummary(itemId) ! json
    case Method.Get -> Root / "item" / Integer(itemId) / "ingredient" / "summary" => 
      itemIngredientSummary(itemId) ! json
    case Method.Get -> Root / "item" / Integer(itemId) / "ingredient" / "tree" => 
      itemIngredientTree(itemId) ! json
  }
```

I'm really not sure I like the new approach better (which in my understanding would like like the following):
```scala
  val routes = (Get / "item" / int / "recipe" / "summary" /> itemRecipeSummary(_) ! json) |
                     (Get / "item" / int / "ingredient" / "summary" /> itemIngredientSummary(_) ! json) |
                     (Get / "item" / int / "ingredient" / "tree" /> itemIngredientTree(_) ! json)
```

One of the things that originally appealed to me with finch is that the routing system just made sense with a basic understanding of scala, it wasn't complicated with a bunch of special operators that I had to learn.  Also, given the fact that routes are defined as partial functions it makes it harder (in my opinion) to mess things up.

Before I chose to use finch the first time I looked at spary and had used unfiltered on a small project.  One of the things I hated with all the fancy syntax of these projects is how easy it was to create code that had bugs in it.

Take for example the following and see if you can spot the bug, mind you this code would compile cleanly:
```scala
def routes = {
  (Get / "user" / int /> user(_) ! json) |
  (Get / "item" / int / "details" /> userDetails(_) ! json)
  (Get / "item" / int / "groups" /> userGroups(_) ! json)
}
```

The bug in the previous example is that I forgot the `|` after the second route definition.  If not paying close attention it's something that is easy to miss and could result is wasted time debugging.

Admittedly, I'm a bit of a noob when it comes to designing libraries in Scala, but the idea of being able to continue using partial functions is something that I would very much like to see stay in the core design of finch.

That said, I think it's probably a good idea to decouple things from the netty and httpx packages.

BenWhitehead commented on core/src/test/scala/io/finch/route/RouterSpec.scala:
```diff
@@ -0,0 +1,172 @@
+    val r4 = Get / string("name") / int("id") / boolean("flag") / "foo"
+    val r5 = Post / * / $
+
+    r1.toString shouldBe "GET "
```
I don't think `GET` is a valid route for an http server, I think the minimum acceptable would be `GET /` which means we probably shouldn't be able to declare a route that consists of only a method.

vkostyukov commented:
Thanks for the feedback @BenWhitehead! And special thanks for the example of the bug. I also like pattern matching on routes: it's simple and easy to understand. But it also a big lack of flexibility that we can just see by the number tickets created against it (copy from #146):

1. We no longer need a type alias for route: #114 
2. We may decouple HttpRequest from Route
3. We would be able to share the sub-route with other routes via composition
4. We finally can define a proper reverse routes
5. We can fetch the path of concrete route which was asked recently on Gitter.

Unfortunately, we won't be able to resolve all of this w/o combinators.

However, I totally agree with you guys @rodrigopr and @BenWhitehead that we have to keep things simple. Finch is a super simple tool so far (I got a feedback recently from a person who wrote a Finch application with zero knowledge of Scala and Finagle, which is the best feedback that I've got so far). So, I would love to see how can we use pattern matching on combinators.

So, I would suggest the following plan:

1. We merge this for now
2. @rodrigopr or I will introduce a PR that allows to use pattern matching over combinators so we can deprecate the old-style `Endpoint` in 0.5.0 release.

What do you say?

BenWhitehead commented:
I think that sounds like a reasonable approach. For me, something almost important to me as simplicity is that the library helps me to write correct code and make it difficult to write incorrect code.

vkostyukov commented:
Thus, we should probably end up having something similar to #79, which was planned as just a fancy version of the current approach.

Regarding the `Get` route. The thing is that `Get` itself is a valid route, but it's string representation looks weird. And I don't really see a simple way to fix it now. So, I think we don't really need to have a string representation that looks exactly like in framework X, but we can try to make it looks similar, which means that we probably want to make it simple. I thing string representation of the router like `GET/:name/:id` looks pretty reasonable for now.

Regarding the `Get` router semantic. The idea is following:

```scala
val r1 = Get // match all the GET requests
val r2 = Get / $ // match GET requests with empty path, like GET /
```

You can still expose any constraints on your routes.

vkostyukov commented:
I thing it's ready to be merged. Last call. Let me know if have any concerns left. Especially about `/>` operator for map. Although, we can easily rename it later.

vkostyukov commented:
Btw, here is a valid code that enables an HTTP echo server (with routing) in Finch:

```scala
Httpx.serve(":8080", Get / "echo" / string /> Ok(_).toFuture)
```

rodrigopr commented:
I'll play a bit with it later this week, see if I can make the pattern-matching work with the new route combinators.
For this PR, I think it's ready to be merged.

vkostyukov commented:
I'm pretty sure we will figure something out. I will also have a look on it this weekend. Now I'm planing to merge this PR and publish a snapshot version.
